nifpga/bitfile.py

```diff
@@ -75,27 +84,480 @@ def base_address_on_device(self):
+        OrderedDict in the correct order"""
+        child = next(child_iter, None)
+        if child is None:
+            return
+        current_result = child.unpack_data(data)
+        data >>= child.size_in_bits
+        self._unpack_data_recursive(data, result, child_iter)
+        result[child.name] = current_result
+
+    def unpack_data(self, data):
+        result = OrderedDict()
+        self._unpack_data_recursive(data, result, reversed(self._children))
+        return result
+
+    def pack_data(self, data_to_pack, packed_data):
+        i = 0
+        for child in self._children:
+            packed_data = child.pack_data(data_to_pack[child.name], packed_data)
+            i += 1
+        return packed_data
+
+
+class _Array(object):
+    """ Handles packing and unpacking arrays. """
+    def __init__(self, name, type_xml):
+        self._name = name
+        self._subtype = _parse_type(list(type_xml.find("Type"))[0])
+        self._size = int(type_xml.find("Size").text)
+        self._size_in_bits = self._subtype.size_in_bits * self._size
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def datatype(self):
+        return self._subtype.datatype
+
+    @property
+    def size(self):
+        return self._size
+
+    @property
+    def size_in_bits(self):
+        return self._size_in_bits
+
+    @property
+    def is_c_api_type(self):
+        return self._subtype.is_c_api_type
+
+    def unpack_data(self, data):
+        results = [0] * self._size
+        for i in range(0, self._size):
+            results[i] = self._subtype.unpack_data(data)
+            data = data >> self._subtype.size_in_bits
+        # Arrays are packed in order, which means that as we are grabbing out values
+        # and shifting data, we are grabbing from the back of the array.  So reverse it.
+        results.reverse()
+        return results
+
+    def pack_data(self, data_to_pack, packed_data):
+        for i in range(0, self._size):
+            packed_data = self._subtype.pack_data(data_to_pack[i], packed_data)
+        return packed_data
+
+
+class _FXP(object):
+    """ Handles packing and unpacking FXP values from the FPGA. """
+    def __init__(self, name, type_xml):
+        self._name = name
+        self._datatype = DataType.Fxp
+        self._signed = True if type_xml.find("Signed").text.lower() == 'true' else False
+        overflow_enabled_xml = type_xml.find("IncludeOverflowStatus")
+        if overflow_enabled_xml is not None:
+            self._overflow_enabled = True if overflow_enabled_xml.text.lower() == 'true' else False
+        else:
+            self._overflow_enabled = False
+        self._word_length = int(type_xml.find("WordLength").text)
+        self._integer_word_length = int(type_xml.find("IntegerWordLength").text)
+        self._delta = self._calculate_delta()
+        self._minimum = self._calculate_minimum()
+        self._maximum = self._calculate_maximum()
```

Comment:
Should put a comment here to why we are calculating these values instead of using the XML version, since the xml value can be coerced